challenges/security/application-security.md

```diff
@@ -0,0 +1,245 @@
+# Summary
+
+We use a two part hiring challenge for application security engineers:
+
+Part One: Peer review. You'll receive a brief Request for Discussion (RFD) and
+architecture diagram of a service similar to docker. You have two hours to provide
+written PR feedback.
+
+Part Two: Development. You'll receive prototype code for the service reviewed
+in part one. Building off this prototype codebase, implement authentication and
+authorization.
+
+# Rationale
+
+These exercises have several goals:
+
+* It helps us to understand what to expect from you as a developer, what security
+  issues you find important, how you communicate those issues to peers and how you
+  write production code.
+* It helps you get a feel for what it would be like to work at Teleport, as this
+  exercise aims to simulate our day-as-usual and expose you to the type of work
+  we're doing here.
+* We aim to keep the process light. When there are time limits, it is to keep
+  exercises from consuming undue amounts of your time, and not to test performance
+  under pressure.
+
+We believe this technique is not only better, but also is more fun compared to
+whiteboard/quiz interviews so common in the industry.
+
+[Some of the best teams use coding challenges.](https://sockpuppet.org/blog/2015/03/06/the-hiring-post/)
+
+We appreciate your time and are looking forward to hack on this project together.
+
+# Leveling
+There are 6 engineering levels at Teleport. It's possible to earn level 4-5 via
+this interview process.
+
+We are not currently hiring L1-L3 security engineers and level 6 is only available
+via internal promotion. See the [engineering levels document](https://raw.githubusercontent.com/gravitational/careers/main/levels.pdf)
+for more details.
+
+We expect different performance at different engineering levels. These expectations
+are explicitly documented below.
+
+
+# Part 1: Peer Review
+You will be given a GitHub repo containing a PR that includes a RFD and architecture
+diagram for a service similar to docker.
+
+You will have two hours to read through the RFD and provide feedback. Please
+focus on security concerns, though you're welcome to provide input in any area
+where you see room for improvement.
+
+## Level 4
+Review the service as written, assuming it will run on a single server.
+
+## Level 5
+The service will run in multiple geographies with unreliable network connections
+between them. Please address the security and business tradeoffs between
+authentication and authorization consistency architectures.
+
+
+# Part 2: Development
+You will be given a prototype implementation of the RFD reviewed in part 1.
```

Comment:
> Can I see the prototype code?

What is this RFD and architecture diagram?

I haven't written it yet -- I'll be working on it today. Notably excluded will be authn/authz, transport security, and logging.